lib/ansible/modules/cloud/vmware/vmware_export_ovf.py

```diff
@@ -0,0 +1,332 @@
+    username: '{{ vcenter_username }}'
+    password: '{{ vcenter_password }}'
+    name: '{{ vm_name }}'
+    export_with_images: true
+    export_dir: /path/to/ovf_template/
+  delegate_to: localhost
+'''
+
+RETURN = r'''
+instance:
+    description: list of the exported files, if exported from vCenter server, device file is not named with vm name
+    returned: always
+    type: dict
+    sample: None
+'''
+
+import os
+import hashlib
+from time import sleep
+from threading import Thread
+try:
+    import requests
+except ImportError:
+    REQUESTS_FOUND = False
+else:
+    REQUESTS_FOUND = True
+
+from ansible.module_utils.basic import AnsibleModule
+from ansible.module_utils._text import to_text
+from ansible.module_utils.vmware import (connect_to_api, vmware_argument_spec, PyVmomi)
+try:
+    from pyVmomi import vmodl, vim
+    from pyVim import connect
+    HAS_PYVMOMI = True
+except ImportError:
+    pass
+
+
+class LeaseProgressUpdater(Thread):
+    def __init__(self, http_nfc_lease, update_interval):
+        Thread.__init__(self)
+        self._running = True
+        self.httpNfcLease = http_nfc_lease
+        self.updateInterval = update_interval
+        self.progressPercent = 0
+
+    def set_progress_percent(self, progress_percent):
+        self.progressPercent = progress_percent
+
+    def stop(self):
+        self._running = False
+
+    def run(self):
+        while self._running:
+            try:
+                if self.httpNfcLease.state == vim.HttpNfcLease.State.done:
+                    return
+                self.httpNfcLease.HttpNfcLeaseProgress(self.progressPercent)
+                sleep(self.updateInterval)
+            except Exception:
+                return
+
+
+class VMwareExportVmOvf(object):
+    def __init__(self, module, vm):
+        self.content = connect_to_api(module)
+        self.module = module
+        self.vm = vm
+        self.params = module.params
+        self.mf_file = ''
+        self.ovf_dir = ''
+        # set read device content chunk size to 2 MB
+        self.chunk_size = 2 * 2 ** 20
+        # set lease progress update interval to 30 seconds
+        self.lease_interval = 30
+        self.facts = {'device_files': []}
+
+    def check_working_dir(self):
+        if 'export_dir' not in self.params or self.params['export_dir'] is None:
+            self.module.fail_json(msg='The path of folder to place exported files not specified.')
+        else:
+            self.ovf_dir = os.path.join(self.params['export_dir'], self.vm.name)
+            if not os.path.exists(self.ovf_dir):
+                try:
+                    os.makedirs(self.ovf_dir)
+                except Exception as err:
+                    self.module.fail_json(msg='Exception caught when create folder %s, with error %s'
+                                              % (self.ovf_dir, to_text(err)))
+        self.mf_file = os.path.join(self.ovf_dir, self.vm.name + '.mf')
+
+    def check_esxi_version(self):
```

Comment:
Perhaps:
```suggestion
    def esxi_version_at_least(self, version):
    """
    Check that the ESXi Host is at least a specific version number.
    
    Inputs:
      - version (tuple): a version tuple, for example (6, 7, 1)
      
     Returns:
       bool
    """
```
Suggest moving into module_utils/vmware.py and providing a unit test.